### PR TITLE
[Enhancement] StoppingCriteria and LogitsProcessor

### PR DESCRIPTION
**Pull request by zhang-ge-hao**
Add functions similar to huggingface's [StoppingCriteria](https://huggingface.co/docs/transformers/internal/generation_utils#transformers.StoppingCriteria) and [LogitsProcessor](https://huggingface.co/docs/transformers/internal/generation_utils#logitsprocessor).

### CODE REVIEW

**Comment by juney-nvidia** [comment]
@zhang-ge-hao 
Thanks for submitting the MR and really appreciate your contributions to TensorRT-LLM.
Our engineers are working on the refinement of our runtime API to make it better.
We plan to integrate your contributions as part of our refinement work and when the work gets landed into the github, we will add you as the co-author and also acknowledge your efforts.

Thanks for the nice contributions. Looking forward to the future work. 

June

**Comment by JohnGiorgi** [comment]
Is there a timeline for this? Asking because it's important for our use case. Thanks a lot for considering adding this functionality!

**Comment by QiJune** [comment]
Sorry for the late reply. We are working on new high level runtime API these days, it has been pulished to the github main branch:  https://github.com/NVIDIA/TensorRT-LLM/blob/main/tensorrt_llm/runtime/model_runner.py, which provides more huggingface style `generate` API.

We will add your enhancement of StoppingCriteria and LogitsProcessor based on the new ModelRunner API, it's expected to be published to the github main branch in the next week.

**Comment by zhang-ge-hao** [comment]
@QiJune @juney-nvidia 

Hi, I reproduced this modification on your newest Python API, with more context passed, and some other adjustments.

You can check #608. 

It's my pleasure if I can provide you with some reference.

**Comment by QiJune** [comment]
@zhang-ge-hao @JohnGiorgi  The `StoppingCriteria` and `LogitsProcessor` base class has been updated to the main branch. Users could develop customized stopping criteria and logits processor.

**Comment by JohnGiorgi** [comment]
Thanks for merging this @QiJune @zhang-ge-hao! Is there any documentation about how to configure the new logits processor in TensorRT-LLM?

**Comment by leopra** [comment]
What about the C++ runtime? Is there any planning to support the stopping criteria?
